### DISCUSSION TITLE
finding-the-urls-of-an-aspnetcore-app-from-a-hosted-service-in-dotnet-6/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#84] [Announcements]
# finding-the-urls-of-an-aspnetcore-app-from-a-hosted-service-in-dotnet-6/

In this post I describe how you can wait for your ASP.NET Core app to be ready to receive requests from inside an IHostedService/BackgroundService in .NET 6

https://andrewlock.net/finding-the-urls-of-an-aspnetcore-app-from-a-hosted-service-in-dotnet-6/

**Comment by nZeus** [comment]
Very nice article!
I have a question: is it possible that `_lifetime.ApplicationStarted.Register()` happens after the application has started?

**Comment by TTWill** [comment]
This article have saved me from many hours of research into working with Kestrel and Windows Service architecture. For some reason, no matter what I did, the service would stop after a few seconds, although other services reported the kestrel was actively responding, but shut down unexpectedly. This article has helped me better understand the middleware architecture and now also solving my problem.

**Comment by alanjh** [comment]
This doesn't totally work as is for IHostedservice, since you can't await in StartAsync as other services won't start until StartAsync is completed. You need to either do no work in StartAsync or wrap it in a Task.Run.

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 15 2022, 08:57, in reply to [Michael Croes](https://disqus.com/by/disqus_NudFLH4CKG/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/84#discussioncomment-6062951)</em>

---

<p>Yep, I think so, thanks. I'll update the code, thanks 🙂</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 15 2022, 08:59, in reply to [Eluvatar](https://disqus.com/by/eluvatar/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/84#discussioncomment-6062956)</em>

---

<p>No, that's very different. Task.FromCompleted creates a task that's _already_ cancelled, and sets the token that did this to the token you pass in. It doesn't create a task that "waits" for the token.</p>

[a comment by an automated account is not shown]

**Comment by careless6666** [comment]
```
  public interface IHostApplicationLifetime
  {
    /// <summary>
    /// Triggered when the application host has fully started.
    /// </summary>
    CancellationToken ApplicationStarted { get; }
...
  }
```

Looks like a slightly strange design from Microsoft. We do a token cancellation, which means that the operation is actually completed and not cancelled. Why not then simply use Task as a sign of completed work, which you can also subscribe to (ContinueWith)?
